scopesim_templates/extragalactic/galaxies.py

```diff
@@ -272,35 +272,37 @@ def spiral_two_component(extent=60*u.arcsec, fluxes=(0, 0), offset=(0, 0)):
 
     filename = "spiral_two_component.fits.bz2"
     path = pathlib.Path(__file__).parent.absolute() / filename
-    hdulist = fits.open(path)
-    img_ext = hdulist[0].header["IMG_EXT"]
-    spec_ext = hdulist[0].header["SPEC_EXT"]
-
-    src = Source()
-    src.fields = hdulist[img_ext:spec_ext]
-    src.spectra = [hdu_to_synphot(hdu) for hdu in hdulist[spec_ext:]]
-
-    for field in src.fields:
-        w, h = field.data.shape
-        field.header["CRPIX1"] = (w + 1) / 2,
-        field.header["CRPIX2"] = (h + 1) / 2,
-        field.header["CRVAL1"] = 0
-        field.header["CRVAL2"] = 0
-        field.header["CDELT1"] = extent / w
-        field.header["CDELT2"] = extent / w
-        field.header["CUNIT1"] = "deg"
-        field.header["CUNIT2"] = "deg"
-        field.header["CTYPE1"] = "RA---TAN"
-        field.header["CTYPE2"] = "DEC--TAN"
-        field.header["SPEC_REF"] = field.header["SPEC_EXT"] - spec_ext
+    with fits.open(path) as hdulist:
+        img_ext = hdulist[0].header["IMG_EXT"]
+        spec_ext = hdulist[0].header["SPEC_EXT"]
+
+        srcs = []
+        for i in range(img_ext, spec_ext):
+            w, h = hdulist[i].data.shape
+            hdulist[i].header["CRPIX1"] = (w + 1) / 2,
+            hdulist[i].header["CRPIX2"] = (h + 1) / 2,
+            hdulist[i].header["CRVAL1"] = 0
+            hdulist[i].header["CRVAL2"] = 0
+            hdulist[i].header["CDELT1"] = extent / w
+            hdulist[i].header["CDELT2"] = extent / w
+            hdulist[i].header["CUNIT1"] = "deg"
+            hdulist[i].header["CUNIT2"] = "deg"
+            hdulist[i].header["CTYPE1"] = "RA---TAN"
+            hdulist[i].header["CTYPE2"] = "DEC--TAN"
+            # hdulist[i].header["SPEC_REF"] = hdulist[i].header["SPEC_EXT"] - spec_ext
+
+            i_spec_ext = spec_ext - img_ext + i
+            srcs.append(Source(image_hdu=hdulist[i],
+                        spectra=[hdu_to_synphot(hdulist[i_spec_ext])]))
+    src = sum(srcs[1:], start=srcs[0])
```

Comment:
While this code is now making better use of the ScopeSim Source interface, it is still rather ugly. Having nice(r) code here is important, because this code also doubles as 'this is a good way to create a Source from an image'-documentation.

In particular, the `src = Source()` and then `src.append(...)` is way more clear than the `srcs = []`, `srcs.append(...)` and subsequently this ugly `sum()` line.

The `for i in range()` is also a bit ugly. Would it be feasible to replace that with an iterator? E.g. `for img_hdu, spec_hdu in zip(hdulist[img_ext:spec_ext], hdulist[spec_ext:]):`?

Perhaps not worthwhile to make this change now.

Reply:
Yeah, I made this rather hastily because things were broken (in hacking into the source fields directly). Not sure though how flexible we need this code to be, i.e. the structure of this particular fits file will not change, yet we dynamically create the source parts (HDU and spectrum), which seems a bit overkill here. So either we refactor this into a separate function that truly dynamically deals with such fits files and then the spiral two components is basically a wrapper for that. Or we could just say this fits file will never change anyway and hardcode the number of extensions in there...